service/worker/processor.go

```diff
@@ -0,0 +1,179 @@
+		consumer     kafka.Consumer
+		isStarted    int32
+		isStopped    int32
+		shutdownWG   sync.WaitGroup
+		shutdownCh   chan struct{}
+		config       *Config
+		logger       bark.Logger
+	}
+)
+
+func newReplicationTaskProcessor(topic, consumer string, client messaging.Client, config *Config,
+	logger bark.Logger) *replicationTaskProcessor {
+	return &replicationTaskProcessor{
+		topicName:    topic,
+		consumerName: consumer,
+		client:       client,
+		shutdownCh:   make(chan struct{}),
+		config:       config,
+		logger: logger.WithFields(bark.Fields{
+			logging.TagWorkflowComponent: logging.TagValueReplicationTaskProcessorComponent,
+			logging.TagTopicName:         topic,
+			logging.TagConsumerName:      consumer,
+		}),
+	}
+}
+
+func (p *replicationTaskProcessor) Start() error {
+	if !atomic.CompareAndSwapInt32(&p.isStarted, 0, 1) {
+		return nil
+	}
+
+	logging.LogReplicationTaskProcessorStartingEvent(p.logger)
+	consumer, err := p.client.NewConsumer(p.topicName, p.consumerName, p.config.ReplicatorConcurrency)
+	if err != nil {
+		logging.LogReplicationTaskProcessorStartFailedEvent(p.logger, err)
+		return err
+	}
+
+	if err := consumer.Start(); err != nil {
+		logging.LogReplicationTaskProcessorStartFailedEvent(p.logger, err)
+		return err
+	}
+
+	p.consumer = consumer
+	p.shutdownWG.Add(1)
+	go p.processorPump()
+
+	logging.LogReplicationTaskProcessorStartedEvent(p.logger)
+	return nil
+}
+
+func (p *replicationTaskProcessor) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.isStopped, 0, 1) {
+		return
+	}
+
+	logging.LogReplicationTaskProcessorShuttingDownEvent(p.logger)
+	defer logging.LogReplicationTaskProcessorShutdownEvent(p.logger)
+
+	if atomic.LoadInt32(&p.isStarted) == 1 {
+		close(p.shutdownCh)
+	}
+
+	if success := common.AwaitWaitGroup(&p.shutdownWG, time.Minute); !success {
+		logging.LogReplicationTaskProcessorShutdownTimedoutEvent(p.logger)
+	}
+}
+
+func (p *replicationTaskProcessor) processorPump() {
+	defer p.shutdownWG.Done()
+
+	var workerWG sync.WaitGroup
+	for i := 0; i < p.config.ReplicatorConcurrency; i++ {
+		workerWG.Add(1)
+		go p.worker(&workerWG)
+	}
+
+processorPumpLoop:
+	for {
+		select {
+		case <-p.shutdownCh:
+			// Processor is shutting down, close the underlying consumer
+			p.consumer.Stop()
+			break processorPumpLoop
+		}
+	}
+
+	p.logger.Info("Replication task processor pump shutting down.")
+	if success := common.AwaitWaitGroup(&workerWG, 10*time.Second); !success {
+		p.logger.Warn("Replication task processor timed out on worker shutdown.")
+	}
+}
+
+func (p *replicationTaskProcessor) worker(workerWG *sync.WaitGroup) {
+	defer workerWG.Done()
+
+	for {
+		select {
+		case msg, ok := <-p.consumer.Messages():
+			if !ok {
+				p.logger.Info("Worker for replication task processor shutting down.")
+				return // channel closed
+			}
+
+			if task, err := deserialize(msg.Value()); err != nil {
+				p.logger.Errorf("Deserialize Error. Value: %v, Error: %v", string(msg.Value()), err)
```

Comment:
if the DLQ part is working in progress, plz add a TODO here.
Since we should investigate any message which cannot be processed.

same applies to line: 151, 160

Reply:
Added a TODO.  Adding a DLQ is pretty straightforward as the underlying kafka-client library already has support for this.  But currently the idea is to block the consumer when this happens until the fix for such issues are rolled out.